Make Edit a type of a Collection

osteslag opened the pull request:
This set of changes is about making `Edit` a type of a *collection* instead of a collection’s *equatable elements* plus the resulting changed requirements, including tests and test app. Merges into the version 3 branch.

An `Edit` needs to have a concept of the collection type so we can ensure we can do arithmetic on the collection’s index distance type, see [discussion](https://github.com/osteslag/Changeset/pull/39#discussion_r129030599) on #39.

I wasn’t able to declare `…where C.IndexDistance: SignedInteger` on `Edit`, `Changeset` while doing enumerations and additions/subtractions on offsets in a compatible way. At least I couldn’t figure out how to. So for now, we require `Int` index distances on the collections.

Note: I prefer to ensure all commits are compilable (`git-bisect`), hence all files in one commit.

FYI: I’m planning a general cleanup for version 3 (including using the `C` shorthand for `Collection` types, probably splitting up `Edit` and `Changeset` in separate files).

osteslag committed:
Make Edit a type of a Collection

An Edit needs to have a concept of the collection type so we can ensure
we can do arithmetic on the collections index distance type.

ole commented:
That was fast! I'll take a look at it soon.

osteslag commented:
I have a full-time job, so I could have been much faster 😉
But compared to the previous PR, based on your comment in April, I guess you’re right 😆

ole commented:
@osteslag I haven't reviewed the changes yet, but I spent some time on getting rid of the `IndexDistance == Int` constraint. See #41.

ole committed:
Remove redundant generic constraints

The T: Collection constraint is redundant because T is already
constrained to Collection through Edit<T> and Swift can infer this.

While the explicit constraint arguably makes some method signatures
clearer, I removed them because Xcode 9 generated warnings.

ole commented:
I pushed one commit removing some redundant constraint. Xcode 9 detected them for me. Xcode 8 doesn't show the warnings, but they are still redundant; the project compiles and runs without them just fine.

<img width="1680" alt="screen shot 2017-07-27 at 15 20 11" src="https://user-images.githubusercontent.com/5010/28672160-420c8cee-72df-11e7-952d-53b07a66e42d.png">

Everything else looks good to me. 👍

osteslag commented:
Good change! Clearly `Collection` as well as the other `Edit` collection constraints are inferred via `Edit<T>`.
